urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/grinder/client/user/Browser.java

```diff
@@ -0,0 +1,242 @@
+import java.net.Socket;
+import java.net.URI;
+import java.net.URL;
+import java.util.UUID;
+
+import websocket.client.DistripediaClient;
+import websocket.client.DistripediaClient.ArticleReceivedEvent;
+import websocket.client.DistripediaClient.ArticleReceivedHandler;
+
+/**
+ * Simulating a very stripped "browser". After initialized it can retrieve
+ * distripedia webpages via a websocket connection or a regular http request.
+ * The fallback to a regular http-request only counts the data received.
+ *
+ * @author Friedemann
+ */
+public class Browser implements Runnable, ArticleReceivedHandler {
+
+	/** @see BrowserType */
+	private final BrowserType type;
+
+	/** The unique browser id. */
+	private final UUID browserID;
+
+	/** stores the http uri */
+	private final URI baseURI;
+
+	/** Stores the websocket uri */
+	private final URI baseURIws;
+
+	/** The local storage object. */
+	private LocalStorage LS = null;
+
+	/** The distripedia client object handling the websocket connections. */
+	private DistripediaClient dpc;
+
+	private Socket stratusSocket;
+
+	private final boolean hasArticleReceived = false;
+
+	private ArticleReceivedEvent articleReceivedEvent = null;
+
+	private final Object waitingForArticleLock = new Object();
+
+	public Browser(final BrowserType type, final URI uriHTTP, final URI uriWS) {
+		browserID = UUID.randomUUID();
+		this.type = type;
+
+		baseURI = uriHTTP;
+		baseURIws = uriWS;
+
+		switch (type)
+		{
+		case WEBSOCKET:
+			LS = new LocalStorage();
+			dpc = new DistripediaClient(uriWS, LS);
+			dpc.setCachedArticlesPreferred(true);
+			dpc.addArticleReceivedHandler(this);
+			break;
+
+		case STRATUS:
+			LS = new LocalStorage();
+			break;
+		}
+	}
+
+	/**
+	 * Gets the browser id.
+	 *
+	 * @return the browser id
+	 */
+	public UUID getBrowserID() {
+		return browserID;
+	}
+
+	/**
+	 * Gets the browser type.
+	 *
+	 * @return BrowserType the browser type
+	 * @see BrowserType
+	 */
+	public BrowserType getBrowserType() {
+		return type;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see
+	 * websocket.client.DistripediaClient.ArticleReceivedHandler#onArticleReceived
+	 * (websocket.client.DistripediaClient.ArticleReceivedEvent)
+	 */
+	@Override
+	public void onArticleReceived(final ArticleReceivedEvent e) {
+		articleReceivedEvent = e;
+
+		// new (david)
+		synchronized (waitingForArticleLock)
+		{
+			waitingForArticleLock.notify();
+		}
+		// end new
+	}
+
+	public ArticleRequest retrieveArticle(final String articleID) {
+
+		final ArticleRequest ar = new ArticleRequest();
+
+		switch (type)
+		{
+		case HTTP:
+
+			retrieveArticleViaHTTP(ar, baseURI.toString() + articleID + ".html");
+			break;
+
+		case WEBSOCKET:
+			articleReceivedEvent = null;
+
+			dpc.getArticle(articleID);
+
+			// new (david)
+			try
+			{
+				synchronized (waitingForArticleLock)
+				{
+					waitingForArticleLock
+							.wait(GrinderSettings.CLIENT_WEBSOCKET_RECEIVE_TIMEOUT);
+				}
+			} catch (final InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)